Add coupled tasks to workflow and fix dependencies

WalterKolczynski-NOAA opened the pull request:
Adds remaining coupled jobs to the free forecast workflow, and additionally addresses a number of other issues out of convenience during the process.

There are two new jobs: one to stage coupled ICs for the prototype, and one to run ocean/ice post. The initialization jobs are now controlled by the APP setting: if the app is S2S or S2SW, the coupled IC job will run in place of the getic and init jobs. Ocean/ice post is a metatask that pools forecast hours in the same manner as atm post, with the number of tasks controlled by the setting in config.ocnpost. These new jobs are only added for the forecast-only workflow.

As part of the addition, a couple of changes were made to how the model output is handled. First, symbolic links are now set up for each of the ocean and ice output files that point to the location on COM, in the same way that is done for atm and wave output, rather than copying the data when the forecast is complete. This allows the data to become available immediately and post to begin while the forecast is still running. Additionally, the output location on COM is moved to component files in the same manned as atm and wave. The joint ocn-ice files are currently located in the ocean directory; a final disposition of these files is under discussion.

While the ocean post jobs were being created, the method of naming the tasks is updated to include the first and last forecast hour processed by the task, rather than a plain index. This change was extended to the regular atm post, including for the full cycled workflow.

To fix the wave post dependencies for coupled mode, two changes were made. First, the wave post data dependency is updated from being hard-coded to the GFS v16 wave grids to reading the wave grid names from waveGRD. Second, a new setting, DOBNDPNT_WAVE, is brought over from coupled-crow to turn off the boundary point post jobs. Since these are changes to existing jobs, the changes were made to the full cycled workflow as well (but need to be tested).

Closes #191, #367, #461, #481, #482, #483, #484, #485, #486, #487

WalterKolczynski-NOAA committed:
Add scripts for staging coupled ICs

Copied script for staging coupled prototype ICs over from coupled-
crow. Existing config script renamed to be more appropriate.

Refs: #367, #481

WalterKolczynski-NOAA committed:
Load NCL for ocean post

Ocean post uses NCL scripts, so an NCL module must be loaded at runtime.

Refs: #483

WalterKolczynski-NOAA committed:
Symlink ocean/wave output to component directories in COM

Ocean and ice output now have symlinks created before the forecast
is run so output is available in COM immediately instead of being
copied when the forecast is complete.

Ocean and ice output in COM are now also located in the appropriate
component directories like data for other components. For now, the
joint ocnice files are placed in the ocean directory. There are
ongoing discussions to determine how these will be handled in the
future.

Refs: #191, #483, #485

WalterKolczynski-NOAA committed:
Change reg2grb2 fix variable to match existing

The name of the variable for the reg2grb2 fix directory had not been
updated from what was used in coupled-crow. The variable name now
matches what is being used in config.base.

Refs: #483

WalterKolczynski-NOAA committed:
Remove redirection of ocnpost output

The output of the NCL scripts run by ocean post was being directed
into files in the working directory, so the output did not appear
in the log. Removing the redirection allows the output to be written
to the log.

Refs: #483

WalterKolczynski-NOAA committed:
Add memory specification for ocean post

Due to a configuration change on RDHPCS machines, jobs no longer
request all core memory by default. This causes performance issues
with some memory-intensive jobs. By making a memory request, we
ensure we have enough memory to complete in the requested wallclock.

Refs: #461

WalterKolczynski-NOAA committed:
Change taskname of post jobs

The task name of post jobs included just an index of the group, so
it was not readily visible what forecast hours are covered by each
task. These task names are now updated to display the first and
last forecast hours processed by the task.

Closes: #484

WalterKolczynski-NOAA committed:
Add ability to turn off wave boundary point post jobs

Adds a setting, DOBNDPNT_WAVE, that allows the boundary point wave
post jobs to be turned off. The setting defaults to YES, but is NO
for the coupled defaults.

In order to make sure the wave config is included in the settings
dict for setup_workflow, workflow utils now includes the config.wave
file in the settings for all jobs that begin 'wave'. This allows the
retrieval of the DOBNDPNT_WAVE setting.

Additionally, the task definition for the gdas and gfs versions of
wavepostpnt is consolidated into a single block to eliminate the
redundant code.

Closes: #486

WalterKolczynski-NOAA committed:
Determine wavepost dependencies from wave grids

Instead of hard-coding the data dependency for the wave grids used
by GFS v16, the workflow now creates the dependency based on waveGRD.

Closes: #487

WalterKolczynski-NOAA committed:
Add coupled tasks to forecast-only workflow setup

Tasks to stage coupled ICs and to run ocean post are added for the
forecast-only workflow setup.

The init jobs run is controlled by the app setting. If the app is
S2S or S2SW, the coupled IC job will run instead of getic and init.
The source locations for the coupled IC job are determined by the
settings in the config.coupled_ic file. There is a base location,
ORIGIN_ROOT, then the subdirectory for each component is given by a
different setting.

Ocean post is pooled into tasks in the same manner as atm post, with
the number of tasks specified in config.ocnpost.

The ocean post config file was missing a line to source the job
resources, which is needed for the workflow generation, so this was
updated.

Refs: #481, #482, #483

WalterKolczynski-NOAA committed:
Remove redundant setting for coupled ICs

There was an unnecessary change in variable name for the base IC
location between the config file and the script. The redundant
ORIGIN_ROOT is replaced with BASE_CPLIC in the script.

Refs: #481

WalterKolczynski-NOAA committed:
Fix permissions on setup scripts

Two of the setup scripts inadvertently had their permissions
changed.

WalterKolczynski-NOAA commented:
A full coupled experiment was tested on Orion. A workflow was generated and manually inspected but not run for atm-only. I'm more concerned about the cycling being bugged by these changes, but cycled is not yet confirmed to work from this branch yet, so that testing is deferred. In particular, need to make sure the analysis job is picked up now that the FHRGRP should be anl instead of 0.

aerorahul commented on jobs/rocoto/coupled_ic.sh:
```diff
@@ -0,0 +1,111 @@
+[[ $status -ne 0 ]] && exit $status
+err=0
+
+export DATAROOT="$RUNDIR/$CDATE/$CDUMP"
```
Do you need `export` here?

WalterKolczynski-NOAA replied:
Actually, we don't need this at all. The directory is created and then never used.

aerorahul commented on jobs/rocoto/ocnpost.sh:
```diff
@@ -124,25 +121,25 @@ for fhr in $fhrlst; do
+      echo "File $COMOUT/ocean/ocn_2D_$VDATE.$ENSMEM.$IDATE.nc already exists"
     else
-      $ncks -x -v vo,uo,so,temp $COMOUT/ocn$VDATE.$ENSMEM.$IDATE.nc $COMOUT/ocn_2D_$VDATE.$ENSMEM.$IDATE.nc
+      ncks -x -v vo,uo,so,temp $COMOUT/ocean/ocn$VDATE.$ENSMEM.$IDATE.nc $COMOUT/ocean/ocn_2D_$VDATE.$ENSMEM.$IDATE.nc
```
can we do a status check here and everywhere we perform this `ncks` business?
```
err=$?
[[ $err-ne 0 ]] && exit $err
```

WalterKolczynski-NOAA replied:
Agree this should happen. This wasn't something I changed.

WalterKolczynski-NOAA replied:
Fixed

aerorahul commented on jobs/rocoto/ocnpost.sh:
```diff
@@ -89,13 +89,10 @@ else
     export COMOUT="$ROTDIR/$CDUMP.$PDY/$cyc"
 fi
-[[ ! -d $COMOUT ]] && mkdir -m 775 -p $COMOUT
+[[ ! -d $COMOUT/ocean ]] && mkdir -m 775 -p $COMOUT/ocean
+[[ ! -d $COMOUT/ice ]] && mkdir -m 775 -p $COMOUT/ice
```
why is the `-m 755` necessary?

WalterKolczynski-NOAA replied:
I was just following what was here before.

aerorahul replied:
I suggest removing it.

WalterKolczynski-NOAA replied:
Going to wait for @JessicaMeixner-NOAA to chime in just in case the more open permissions are needed for the prototype work. Even if it is needed, this is probably not the best ones to use. For reference, atmos has the following permissions: `rwxr-s---`, compared to the `rwxrwsr-x` for the other directories (wave also uses these broader permissions).

JessicaMeixner-NOAA replied:
I don't know of an issue/reason for needing this. Thanks for checking w/me.

WalterKolczynski-NOAA replied:
Done, though FWIW the wave mkdirs also set the permissions, so we'll want to revisit that later.

aerorahul commented on jobs/rocoto/ocnpost.sh:
```diff
@@ -124,25 +121,25 @@ for fhr in $fhrlst; do
-      $ncks -v uo -d yh,503 -d xh,-299.92,60.03 $COMOUT/ocn_3D_$VDATE.$ENSMEM.$IDATE.nc $COMOUT/ocn-uo-EQ_$VDATE.$ENSMEM.$IDATE.nc
+      ncks -v uo -d yh,503 -d xh,-299.92,60.03 $COMOUT/ocean/ocn_3D_$VDATE.$ENSMEM.$IDATE.nc $COMOUT/ocean/ocn-uo-EQ_$VDATE.$ENSMEM.$IDATE.nc
     fi
     status=$?
     [[ $status -ne 0 ]] && exit $status
```
This will only trap tha last execution and is either inadequate or unnecessary if every file is present.

WalterKolczynski-NOAA replied:
Fixed

aerorahul commented on jobs/rocoto/ocnpost.sh:
```diff
@@ -153,7 +150,7 @@ done
 echo $pwd
-$NMV ocn_ice*.grb2 $COMOUT
+$NMV ocn_ice*.grb2 $COMOUT/ocean/
 status=$?
 [[ $status -ne 0 ]] && exit $status
```
we are checking on every CP and MV now?

WalterKolczynski-NOAA replied:
This either.

aerorahul replied:
I leave it to you.  I would suggest removing this instance of status check for move.
OTOH, would creating soft-links help "making these files available faster".
Probably can defer that for a later PR.

WalterKolczynski-NOAA replied:
Now that there is proper exit code checking for all the ncks calls, I'm going to leave this so everything is being checked.

aerorahul commented on jobs/rocoto/ocnpost.sh:
```diff
@@ -153,7 +150,7 @@ done
 # Restore CDATE to what is expected
 export CDATE=$IDATE
 echo $pwd
```
remove.

aerorahul replied:
This line has no meaning.

WalterKolczynski-NOAA replied:
This wasn't something I added.

aerorahul replied:
true.
But you can remove.

WalterKolczynski-NOAA replied:
Removed

aerorahul commented on ush/forecast_postdet.sh:
```diff
@@ -797,57 +786,77 @@ MOM6_out() {
     export COMOUT="$ROTDIR/$CDUMP.$PDY/$cyc"
   fi
-  [[ ! -d $COMOUT ]] && mkdir -m 775 -p $COMOUT
+  [[ ! -d $COMOUT/ocean ]] && mkdir -m 775 -p $COMOUT/ocean
```
`-m 775`, why?

WalterKolczynski-NOAA replied:
Just following what was here before.

aerorahul replied:
remove.

WalterKolczynski-NOAA replied:
done

aerorahul commented on ush/forecast_postdet.sh:
```diff
@@ -916,60 +925,51 @@ CICE_postdet() {
+  # Link output files
+  export ENSMEM=${ENSMEM:-01}
+  export IDATE=$CDATE
+  [[ ! -d $COMOUT/ice ]] && mkdir -m 775 -p $COMOUT/ice
```
`-m 775`, why?

WalterKolczynski-NOAA replied:
As before.

aerorahul replied:
remove.

WalterKolczynski-NOAA replied:
done

aerorahul commented on ush/forecast_postdet.sh:
```diff
@@ -797,57 +786,77 @@ MOM6_out() {
+  if [ ${inistep:-"restart"} = 'cold' ]; then
+    $NLN $COMOUT/ocean/ufs.cpld.cold.cpl.r.*.nc $DATA/ufs.cpld.cold.cpl.r.*.nc
+    $NLN $COMOUT/ocean/rpointer.cpl $DATA/rpointer.cpl
     status=$?
     exit $status
```
Linking a non-existent file will create a dead link.  The status will always be 0, no?

WalterKolczynski-NOAA replied:
I think so. The check is left over from it being a copy. Will remove.

WalterKolczynski-NOAA replied:
I guess it will be non-zero if it fails to create the link. Not sure if that is worth checking though, since if it can't write to $DATA, there are loads of other problems.

WalterKolczynski-NOAA replied:
Fixed

aerorahul commented on ush/forecast_postdet.sh:
```diff
@@ -797,57 +786,77 @@ MOM6_out() {
-
-        echo "$NCP -p $ocnfile $COMOUT/ocn$p_date.$ENSMEM.$IDATE.nc"
-        $NCP -p $ocnfile $COMOUT/ocn$p_date.$ENSMEM.$IDATE.nc
+      if [ $fhr = 'anl' ]; then
```
Suggestion, may be for future if it is too much in this PR.
Can we move this to an e.g. `forecast_postdet_ocean.sh`?

WalterKolczynski-NOAA replied:
I have no problem with that, but we should break the file up for each component to be consistent.

aerorahul reviewed (changes requested):
I made a pass through the shell scripts.  I did not review the python scripts and will do so in a bit.
Few comments:
1. `ncks` commands could be checked for return status.
2. `-m 775` is unnecessary and if it is, then should be done everywhere.
3.  Some minor points.

All-in-all, looking good.  Thanks for doing this "overnight"

aerorahul commented on ush/rocoto/workflow_utils.py:
```diff
@@ -113,6 +113,9 @@ def source_configs(configs, tasks):
         elif task in ['efcs']:
             files.append(find_config('config.fcst', configs))
             files.append(find_config('config.efcs', configs))
+        elif task[0:4] in ['wave']:
```
hmm ... substring?

aerorahul replied:
```suggestion
        elif 'wave' in task:
```

WalterKolczynski-NOAA replied:
Didn't occur to me I could do it this way. My brain clearly still hasn't fully adapted to the pythonic ways.

WalterKolczynski-NOAA replied:
Fixed

aerorahul commented on ush/rocoto/setup_workflow_fcstonly.py:
```diff
@@ -250,55 +254,97 @@ def get_workflow(dict_configs, cdump='gdas'):
+        base_cplic = dict_configs['coupled_ic']['ORIGIN_ROOT']
+
+        # ATM ICs
+        for file in ['gfs_ctrl.nc'] + [f'{a}_data.tile{n}.nc' for a in ['gfs', 'sfc'] for n in range(1, 7)]:
```
I am guilty of this comment, but I have learned that use of single character variables e.g. `a`  and `n` is generally not a good practice.
Going forward, I think we should avoid them.  Replacing `a` with `aa` and `n` with `nn` or something like that does not add too much burden, but helps `grepping` later.

aerorahul replied:
On another note, I assume `S2S` variants are only starting from an atmospheric cold-start.
Perhaps that is known.

WalterKolczynski-NOAA replied:
I generally avoid single-character comments in general unless it is a one-line for loop like this. But I have no problem changing it. I'm pretty sure you are correct on the cold-start, at least for the time being.

WalterKolczynski-NOAA replied:
Replaced with more verbose indicies

aerorahul commented on ush/rocoto/setup_workflow_fcstonly.py:
```diff
@@ -311,8 +357,9 @@ def get_workflow(dict_configs, cdump='gdas'):
         deps.append(rocoto.add_dependency(dep_dict))
-        dep_dict = {'type': 'task', 'name': f'{cdump}init'}
-        deps.append(rocoto.add_dependency(dep_dict))
+        if app not in ['S2S', 'S2SW']:
```
I don't understand this conditional.
Can you please explain what is the objective here?

WalterKolczynski-NOAA replied:
The {cdump}init job does not run for S2S modes, so waveinit can't depend on it. 

I'd love for any future workflow generation utilities to have a validation step that just removes dependencies on non-existent jobs like this.

aerorahul replied:
Ok.

aerorahul reviewed (changes requested):
A coupled of comments and recommendation on the python scripts.

WalterKolczynski-NOAA reviewed (commented):
I knew I wasn't going to be able to sleep before I got it all committed anyway.

Most of the comments seem to be about stuff I didn't actually change, but I have no problem fixing. Probably should be a new PR though, this one is already bloated with too many fixes.

WalterKolczynski-NOAA committed:
Remove unnecessary exit code checks when creating symlinks

After switching from copies to symlinks, there were some unneeded
exit code checks left behind.

Refs: #483, #485

WalterKolczynski-NOAA committed:
Add exit code checks after ncks calls

Ocean post had been checking the exit status only once, after the final
ncks call. This would not capture any errors from previous ncks calls.

Refs: #483

WalterKolczynski-NOAA committed:
Check for wave tasks in more pythonic way

When workflow utils checks if a job is a wave task, it now looks for
the word wave anywhere in the task instead of just looking at the
first four characters.

Refs: #482

aerorahul reviewed (commented):
getting closer.

WalterKolczynski-NOAA committed:
Use more verbose loop indicies

Updates some of the new single-character loop indicies to be more
descriptive of what they represent.

Refs: #488

WalterKolczynski-NOAA committed:
Remove extra permissions from ocean/ice COM directories

The ocean and ice directories in COM were being created with
specific permissions not used for atmos.

Refs: #482, #483, #485

WalterKolczynski-NOAA committed:
Remove unneeded pwd from ocean post

Refs: #483

aerorahul reviewed (approved):
Thanks for addressing the comments.
Code changes look good.
Lets see what the tests reveal.

WalterKolczynski-NOAA committed:
Stop creating unneeded working directory for coupled IC job

A working directory was being created by the coupled IC job, but it
was never used because all the files are copied directly to their
destination.

Refs: #481

WalterKolczynski-NOAA committed:
Remove wavepostbndpnt from arch dependency if it is off

When the ability to turn off wave post for boundary points was
added, the dependency for arch was not updated.

Refs: #486

JessicaMeixner-NOAA reviewed (approved):
@WalterKolczynski-NOAA I don't think I'm going to get a chance to test this today - but all the changes look good to me so I'm going to go ahead and approve.  I'll try to test next week when I get a chance.  Also --- THANK YOU! Splitting the ocn/ and ice/ to their respective com directories is awesome - as is allowing ocean post to run as the model progresses.  Great updates along with just bringing over these jobs.

WalterKolczynski-NOAA committed:
Fix error in workflow generation for coupled IC

When removing the redundnant setting in 47a0973, the setting was
not changed in the workflow setup script.

Refs: #481

WalterKolczynski-NOAA committed:
Export coupled IC location

Export the base location for coupled ICs so workflow utils parses
it.

WalterKolczynski-NOAA committed:
Fix syntax error in ocnpost

A syntax error was introduced when adding the exit code checks for
ncks.

Refs: #483

WalterKolczynski-NOAA commented:
Retested coupled on both orion and hera, and also atm-only free forecast on orion. All looks good.
